### PR TITLE
bpo-42345: Fix three issues with typing.Literal parameters

### PR DESCRIPTION
**Pull request by uriyyo**
Literal equality no longer depends on the order of arguments.

Fix issue related to `typing.Literal` caching by adding `typed` parameter to `typing._tp_cache` function.

Add deduplication of `typing.Literal` arguments. 

https://bugs.python.org/issue42345

Automerge-Triggered-By: GH:gvanrossum

### CODE REVIEW

File: Lib/typing.py
Comment (Fidget-Spinner) [inline Lib/typing.py]: One problem with this:
```python
>>> Literal[1, 2, 3] == Literal[1, 2, 3, 3]
False
```
According to [the pep](https://www.python.org/dev/peps/pep-0586/#shortening-unions-of-literals):
> Literal[v1, v2, v3] is equivalent to Union[Literal[v1], Literal[v2], Literal[v3]]

Union de-duplicates args, so IMO, Literal should do this too. There are two approaches possible for this:
1. Change the return in `__eq__` to just:
```suggestion
        set(self.__args__) == set(other.__args__)
```
This preserves the original `__args__`.

2. Or just return `set(params)` instead of `tuple(params)` in `_flatten_literal_params`.
This doesn't preserve the original `__args__`, but it's how Union handles it. So maybe this is preferrable?

**Review by Fidget-Spinner — COMMENTED** [review]
Hello, thank you for your PR to cpython! This looks very good to me, just one thing about de-duplicating literals pointed out in the review below.

File: Lib/typing.py
Comment (uriyyo) [reply Lib/typing.py]: Thanks for the suggestions.

First option can't be used because `Literal[0] == Literal[False]` will be `True` and peps says:
>  Literal[0] and Literal[False] is not equivalent despite that 0 == False evaluates to 'true' at runtime: 0 has type int and False has type bool.

Regarding the second option, when you trying to create `Literal` with mutable type, for instance `Literal[{1: 2}]`, it should not fail at runtime, so if we will try to create `Literal[{1: 2}]` it will fail because the parameter is not hashable.

I have updated the pull request. Basically, it will construct a set of pairs (parameter and parameter type) for literals and then compare those sets. So code like this `Literal[False] == Literal[0]` will be `False`.

@Fidget-Spinner what is your opinion regarding such solution?

Comment (Fidget-Spinner) [reply Lib/typing.py]: Oops, you're right about both points, I forgot about the special `{0} == {False}`, and that there can be non-hashable args for Literals.

I think this is a good solution, thanks.

File: Lib/test/test_typing.py
Comment (Fidget-Spinner) [inline Lib/test/test_typing.py]: I think we should keep the deleted test, it may be useful so that there isn't an accidental regression by someone else in the future.

**Review by Fidget-Spinner — COMMENTED** [review]
I still think that Literal should de-duplicate its `__args__` similar to Union, although I now see the `set` approach won't work. Maybe you can call `set` on the ` _value_and_type_iter` function, then extract out the parameters again, something like:
```python
def Literal:
    ...
    parameters = [p for p, _ in set(_value_and_type_iter(parameters))]
    return _LiteralGenericAlias(self, parameters)
```
This probably needs more discussion, and Guido/Ivan's decision on what the expected behavior should be. Maybe we can ask on the bpo issue.

File: Lib/test/test_typing.py
Comment (uriyyo) [reply Lib/test/test_typing.py]: Thanks, I will return this test

**Comment by uriyyo** [comment]
@Fidget-Spinner There is a problem with `Literal` caching.

Code like this will fail because of type caching:
```python
>>> Literal[0] == Literal[False]
False
```

Code below will show the root cause of this problem:
```python
>>> (Literal[0], Literal[False])
(Litera[0], Literal[0])
```

As a solution, we can disable caching for a `Literal`.

What do you think about that?

**Comment by Fidget-Spinner** [comment]
> Code like this will fail because of type caching:

One possible workaround for that is to pass `typed=True` for `lru_cache` in `tp_cache`, then it will differentiate 0 and False.

The only problem is I realized the test will still fail for passing in dict Literal.

A possible workaround is to just loop through them and check the other args one by one (not particularly efficient but I don't expect Literals to be too long), something like what `_remove_dups_flatten` for Union does, but instead you can't use `set` due to non-hashable args.

I think before we progress further, we should wait for a core dev to respond on the bpo. They might think just having comparisons work is sufficient, and that de-duplicating args isn't needed. So let's save our effort for now 😉 .

**Comment by uriyyo** [comment]
@Fidget-Spinner I forget about `lru_cache` `typed` parameter, thanks for the reminder

I totally agree with you, let wait for a decision from Guido and Ivan regarding `Literal` args de-duplicating.

As for me, we can skip args de-duplication because `Literal` equals method will convert everything to `set` so duplicates won't be a problem.

**Comment by gvanrossum** [comment]
Please change the title and news text. We don’t use the style that describes the new state, we describe the action of the PR as a change to the behavior. For examples, just look at the commit log.

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py:252]: Add a note that this supports both `@tp_cache` and `@tp_cache(typed=True)`.

(And honestly I think there are nicer ways to write the logic for that. Maybe this works?
```
def _tp_cache(func=None, /, *, typed=False):
    if func is not None:
        return _tp_cache(typed=False)(func)
```
)

**Review by gvanrossum — COMMENTED** [review]
LG, except for one nit.

You might mention in the description (not the title) and the news file that you also fixed an issue where incorrectly `Literal[0] == Literal[False]` -- good catch, that!

File: Lib/typing.py
Comment (uriyyo) [reply Lib/typing.py:252]: You are right, such `_tp_cache` signature looks much cleaner.

There are so many ways to make a decorator with parameters. It will be great to have some kind of decorator at `functools` which will allow us to easily write  decorators with parameters, it might look like this:
```py
from functools import decorator

@decorator
def _tp_cache(func=None, /, *, typed=False):
    """Internal wrapper caching __getitem__ of generic types with a fallback to
    original function for non-hashable arguments.
    """
    cached = functools.lru_cache(typed=typed)(func)
    _cleanups.append(cached.cache_clear)

    @functools.wraps(func)
    def inner(*args, **kwds):
        try:
            return cached(*args, **kwds)
        except TypeError:
            pass  # All real errors (not unhashable args) are raised below.
        return func(*args, **kwds)
    return inner
```

Should I create an issue regarding that?

**Comment by uriyyo** [comment]
@gvanrossum I have seen your message regarding `Literal` args deduplication, so I have implemented it in the scope of this PR.

File: Lib/typing.py
Comment (Fidget-Spinner) [inline Lib/typing.py:981]: `__eq__` will throw error in this scenario if we pass a non-hashable literal because of the outer `set` call: 
```python
from typing import Literal

Literal[{'a': 'b'}] == Literal[{'a': 'b'}]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
    return set(_value_and_type_iter(self.__args__)) == set(_value_and_type_iter(other.__args__))
TypeError: unhashable type: 'dict'
```
A possible workaround is to do 
```python
# The O(N^2) method provided by Raymond from 
# https://stackoverflow.com/questions/7828867/how-to-efficiently-compare-two-unordered-lists-not-sets-in-python
compare(_deduplicate(self.__args__), _deduplicate(other.__args__))
```

Btw: I think there should be a test for that specific scenario above. Like a `self.assertEqual(Literal[{'a':'b'}], Literal[{'a':'b'}])` .

File: Lib/typing.py
Comment (Fidget-Spinner) [inline Lib/typing.py:499]: We can choose not to rely on `set` in `_deduplicate`, otherwise if we use `set`, the following won't deduplicate:
```python
>>> Literal[1,1,2, {'a':'b'}, {'a':'b'}]
typing.Literal[1, 1, 2, {'a': 'b'}, {'a': 'b'}]
```
(I added a suggestion below to work around this for _deduplicate)
This way, we don't need the try and except block.

File: Lib/typing.py
Comment (Fidget-Spinner) [inline Lib/typing.py:207]: The following code will de-duplicate even non-hashable types. It's basically what the union type expression C code's dedup does:
```suggestion
    all_params = [p for index, p in enumerate(params) if p not in params[:index]]
```

**Review by Fidget-Spinner — COMMENTED** [review]
Getting very close :). Thank you for your patience.

File: Lib/typing.py
Comment (uriyyo) [reply Lib/typing.py:981]: I think we can skip this because PEP said that `Literal` can be used with mutable values, we don't raise an exception when creating `Literal` with mutable values but I think compare can raise an exception.

What do you think about that?

Comment (Fidget-Spinner) [reply Lib/typing.py:981]: That's an interesting point. For the most part, `Literal` should be used by static type checkers and not in equality tests, so it makes sense. I'm just afraid someone gets this error in the future and files a bug report.

Comment (uriyyo) [reply Lib/typing.py:981]: As an option, we can catch a `TypeError` and raise a new `TypeErrror` with information about the usage of mutable value.

File: Lib/typing.py
Comment (uriyyo) [reply Lib/typing.py:499]: Looks like all comments are regarding the same problem: should we support mutable when compare `Literals` or not.

File: Lib/typing.py
Comment (uriyyo) [reply Lib/typing.py:207]: This method is used by `Union` and for `Union` all parameters are hashable, so it will affect the performance of `Union`.

In case if we will support mutable values we should create a separate `_deduplicate` function for `Literal` type.

Comment (Fidget-Spinner) [reply Lib/typing.py:207]: You're right, I missed that in the last commit you separated out `_dedup` and re-used for both `Union` and `Literal`. Yes in that scenario we can create a separate function.

File: Lib/typing.py
Comment (gvanrossum) [reply Lib/typing.py:499]: So does the PEP allow mutable literals? IIIRC it only allows a specific list of types.

Comment (uriyyo) [reply Lib/typing.py:499]: Looks like it doesn't.

> Literal may be parameterized with literal ints, byte and unicode strings, bools, Enum values and None. 

Also, PEP said:

> Although the set of parameters Literal[...] may contain at type check time is very small, the actual implementation of typing.Literal will not perform any checks at runtime.

So I think we can skip deduplication and equality for mutable types.

Comment (Fidget-Spinner) [reply Lib/typing.py:499]: From [PEP 586](https://www.python.org/dev/peps/pep-0586/#illegal-parameters-for-literal-at-type-check-time)
>Illegal parameters for Literal at type check time
>Mutable literal data structures like dict literals, list literals, or set literals: literals are always implicitly final and immutable. So, Literal[{"a": "b", "c": "d"}] is illegal.

Let's drop this then, sorry to all.

Edit: In this case, I think it's perfectly fine to not catch the error and just let it bubble up. After all, such usage is illegal.

Comment (uriyyo) [reply Lib/typing.py:499]: @Fidget-Spinner can I resolve all conversation?

Comment (Fidget-Spinner) [reply Lib/typing.py:499]: Go ahead, sorry for this wild goose chase 🙏

Comment (uriyyo) [reply Lib/typing.py:499]: No worries, It was interesting to dig more information regarding `PEP 586`🙂

**Review by Fidget-Spinner — APPROVED** [review]
LGTM, and sorry for the weird tangents I went off on in some places!

**Comment by uriyyo** [comment]
@Fidget-Spinner thanks for the review and all comments 🙂

What will be the next step regarding this PR?

**Comment by Fidget-Spinner** [comment]
> @Fidget-Spinner thanks for the review and all comments 🙂
> 
> What will be the next step regarding this PR?

A core dev like Guido/Ivan will take a look, once they're satisfied and approve it, it will be merged into 3.10. And probably backported to 3.9 and 3.8 too (done automatically via bots). This may take some time depending on how busy the core devs are.

I'm looking forward to see your first commit to cpython :).

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: I just thought of something. Why not avoid the recursion and instead move this to after the definition of 'decorator()', like this:
```py
if func is not None:
    return decorator(func)
```

**Comment by gvanrossum** [comment]
I will look at it some time this week.

File: Lib/typing.py
Comment (Dominik1123) [inline Lib/typing.py:242]: The [`_remove_dups_flatten`](https://github.com/python/cpython/blob/3.9/Lib/typing.py#L212) of `Union` contains this additional check:

```py
elif isinstance(p, tuple) and len(p) > 0 and p[0] is Union:
    params.extend(p[1:])
```

I wonder if such a check is relevant for `Literal` as well?

Comment (uriyyo) [reply Lib/typing.py:242]: Good catch, but I can realize what `Union` declaration case is covered by this if statement.
`Literal` tests cover all possible declaration mentioned at PEP. 

Can you provide `Literal` declaration which won't be handled with the current implementation?

**Comment by Dominik1123** [comment]
Should the docs of [`typing.get_args`](https://docs.python.org/3/library/typing.html#typing.get_args) be updated to mention `Literal` as well (similar to #23254)?

**Comment by gvanrossum** [comment]
> Should the docs of [`typing.get_args`](https://docs.python.org/3/library/typing.html#typing.get_args) be updated to mention `Literal` as well (similar to #23254)?

Sure, but I would make a minimal change, e.g. changing "Union" in that paragraph to "Union or Literal".

File: Lib/typing.py
Comment (gvanrossum) [reply Lib/typing.py:242]: I don't understand why there would be a nested tuple in Union parameters. Can you show an example that ever triggers that check?

File: Lib/test/test_typing.py
Comment (gvanrossum) [inline Lib/test/test_typing.py:531]: Are there checks verifying the value of `__args__`? Should it be `(1, 2, 3)` or `(1, 2, 3, 3)`? I can't easily follow the path through the code that would take. :-(

File: Lib/test/test_typing.py
Comment (gvanrossum) [inline Lib/test/test_typing.py]: Again, maybe some checks verifying `__args__`?

File: Misc/NEWS.d/next/Library/2020-11-15-15-23-34.bpo-42345.hiIR7x.rst
Comment (gvanrossum) [inline Misc/NEWS.d/next/Library/2020-11-15-15-23-34.bpo-42345.hiIR7x.rst]: I would rewrite this as a single sentence, e.g. "Fix various issues with typing.Literal parameter handling (flatten, deduplicate, use type to cache key)."

File: Lib/typing.py
Comment (gvanrossum) [reply Lib/typing.py:499]: So the try/except should be removed?

File: Lib/typing.py
Comment (uriyyo) [reply Lib/typing.py:242]: I have just run `test_typing` and no test didn't enter this if statement.
Looks like it's a dead code that we can remove.

Comment (gvanrossum) [reply Lib/typing.py:242]: Well at least we shouldn’t replicate it for Literal.

File: Misc/NEWS.d/next/Library/2020-11-15-15-23-34.bpo-42345.hiIR7x.rst
Comment (uriyyo) [reply Misc/NEWS.d/next/Library/2020-11-15-15-23-34.bpo-42345.hiIR7x.rst]: Thanks,  it sounds much cleaner.

File: Lib/test/test_typing.py
Comment (uriyyo) [reply Lib/test/test_typing.py]: Added tests to cover `Literal.__args__`

File: Lib/typing.py
Comment (uriyyo) [reply Lib/typing.py]: Makes sense, thanks.

File: Lib/typing.py
Comment (uriyyo) [reply Lib/typing.py:242]: Makes sense, I will resolve this conversation.

**Review by gvanrossum — APPROVED** [review]
Thanks, great work!

**Comment by gvanrossum** [comment]
Hm, I am changing my mind. The bug may exist in 3.8 and 3.9, but the fix is feeling a bit too much like a feature (at least flattening and deduping) that we may not want to backport. What do you think?

@ambv (release manager) any thoughts on this? Note that this fixes three things:
- Bug where `Literal[1] == Literal[True]` (as static types these are different!)
- Dedupe, so `Literal[1, 1] == `Literal[1]`
- Flatten, so `Literal[Literal[1], 2] == Literal[1, 2]`

**Comment by ambv** [comment]
It is a bit late in 3.8 I guess. I'd say, let's leave 3.8 alone but this should definitely go into 3.9.1. That way it's relatively easy to say "well, before 3.9.1 we had this buggy Literal behavior".

**Comment by gvanrossum** [comment]
Okay, @uriyyo, you can try the backport to 3.9. Let me know if you need help!

**Comment by uriyyo** [comment]
@gvanrossum Looks like I did it🙂

**Comment by Dominik1123** [comment]
> > Should the docs of [`typing.get_args`](https://docs.python.org/3/library/typing.html#typing.get_args) be updated to mention `Literal` as well (similar to #23254)?
> 
> Sure, but I would make a minimal change, e.g. changing "Union" in that paragraph to "Union or Literal".

@uriyyo What about updating the docs of `typing.get_args`?

**Comment by Fidget-Spinner** [comment]
> > > Should the docs of [`typing.get_args`](https://docs.python.org/3/library/typing.html#typing.get_args) be updated to mention `Literal` as well (similar to #23254)?
> > 
> > 
> > Sure, but I would make a minimal change, e.g. changing "Union" in that paragraph to "Union or Literal".
> 
> @uriyyo What about updating the docs of `typing.get_args`?

@Dominik1123, not sure if this is a bug, but `hash()` of Literal is dependent on args order in the current code, so the type caching problem won't happen.

```python
>>> hash(Literal[1, 2, 3])
3136223724409791019
>>> hash(Literal[3, 2, 1])
4438910888190981000
```

**Comment by uriyyo** [comment]
@Fidget-Spinner Looks like we should not update docs because `Literal` can't be parametrized with generics so this problem will not exist for `Literal`. Please correct me if I miss smth.

**Comment by gvanrossum** [comment]
But Literal can occur inside other types, e.g. `Tuple[str, str, Literal[0, 1, 2]]`.

**Comment by Fidget-Spinner** [comment]
@uriyyo yeah, sorry if I wasn't clear in my original statement - I was in support of not updating the docs. Although my reasoning is slightly different. It's not about generics, the behavior Dominik mentioned happens when Union's nested in _anything_. The real reason why we don't need a docs update is because the `__hash__` of Literal in your patch is order-dependent.

**Comment by uriyyo** [comment]
@Fidget-Spinner Thanks for the explanation it's clear to me now.

**Comment by Fidget-Spinner** [comment]
@uriyyo Awesome! I'm glad I was of some help :).

**Comment by Dominik1123** [comment]
> > > > Should the docs of [`typing.get_args`](https://docs.python.org/3/library/typing.html#typing.get_args) be updated to mention `Literal` as well (similar to #23254)?
> > > 
> > > 
> > > Sure, but I would make a minimal change, e.g. changing "Union" in that paragraph to "Union or Literal".
> > 
> > 
> > @uriyyo What about updating the docs of `typing.get_args`?
> 
> @Dominik1123, not sure if this is a bug, but `hash()` of Literal is dependent on args order in the current code, so the type caching problem won't happen. Currently, it's still only Union affected.
> 
> ```python
> >>> hash(Literal[1, 2, 3])
> 3136223724409791019
> >>> hash(Literal[3, 2, 1])
> 4438910888190981000
> ```

@Fidget-Spinner @uriyyo I see, I didn't notice before. But isn't this very unusual in terms of behavior? So we can have two objects that compare equal but their hashes are different. AFAIK nowhere else in Python this is the case. The closest that comes to it in terms "being equal, but not having the same hash" is `set` and `frozenset` but in that case at least `set` will raise. Now for `Literal` it's yet another level, since the hashes are actually computed, and it's even the same type.

To quote the [`Python data model`](https://docs.python.org/3/reference/datamodel.html#object.__hash__) on `__hash__`:

> The only required property is that objects which compare equal have the same hash value;

So according to that it's a bug and the hash should not be order dependent.

**Comment by uriyyo** [comment]
@Dominik1123 Agree with you, looks like it's a bug related to `Literal` hash implementation, basically, we should change implementation from:
```python
    def __hash__(self):
        return hash(tuple(_value_and_type_iter(self.__args__)))
```
to:
```python
    def __hash__(self):
        return hash(frozentset(_value_and_type_iter(self.__args__)))
```

@Fidget-Spinner should we create issue regarding this bug?

**Comment by Dominik1123** [comment]
> @Dominik1123 Agree with you, looks like it's a bug related to `Literal` hash implementation, basically, we should change implementation from:
> 
> ```python
>     def __hash__(self):
>         return hash(tuple(_value_and_type_iter(self.__args__)))
> ```
> 
> to:
> 
> ```python
>     def __hash__(self):
>         return hash(frozentset(_value_and_type_iter(self.__args__)))
> ```
> 

I'm not sure if `frozenset` is too restrictive in terms of the values it can work with. Right now [PEP 586](https://www.python.org/dev/peps/pep-0586/#illegal-parameters-for-literal-at-type-check-time) disallows literals of mutable types, but only at type check time. The PEP also contains the following comment:

> [...] the actual implementation of `typing.Literal` will not perform any checks at runtime. [...] This is partly to help us preserve flexibility in case we want to expand the scope of what Literal can be used for in the future [...]

Using `frozenset` would take away on that flexibility regarding literals of mutable types. Not sure if that's relevant though. I think we should wait for @gvanrossum to comment on this.

**Comment by gvanrossum** [comment]
Whoa, this is a bug. It not allowed to have two objects that compare equal but have a different hash. So since `Literal[1, 2] == Literal[2, 1]`, those two should have the same hash. I'm sorry I didn't catch this!

Using set() or frozenset() here is totally fine, we already use those in `__eq__`. I'd use set().

**Comment by gvanrossum** [comment]
Create a new PR pointing to the same issue -- I already reopened it.

**Pull request merged**
